src/languages/tcl.js

```diff
@@ -40,14 +44,13 @@ export default function(hljs) {
       },
       {
         excludeEnd: true,
+        className: "variable",
         variants: [
           {
-            begin: '\\$(\\{)?(::)?[a-zA-Z_]((::)?[a-zA-Z0-9_])*\\(([a-zA-Z0-9_])*\\)',
-            end: '[^a-zA-Z0-9_\\}\\$]'
+            begin: '\\$(::)?[a-zA-Z_]+((::)?[a-zA-Z0-9_]+)*' + regex.optional(ARRAY_ACCESS),
           },
           {
-            begin: '\\$(\\{)?(::)?[a-zA-Z_]((::)?[a-zA-Z0-9_])*',
-            end: '(\\))?[^a-zA-Z0-9_\\}\\$]'
+            begin: '\\$\\{(::)?[a-zA-Z_]((::)?[a-zA-Z0-9_])*' + regex.optional(ARRAY_ACCESS) + '\\}',
```

Comment:
Hmm... This doesn't look right to me. Consider e. g. `$foo($bar)`: array index can be a full blown expression. You can have more fun:
```tcl
set foo bar 
set ${foo}(1) 42 ; # this is not captured by the grammar
puts $bar(1)
```
It has nothing to do with autodection fixes though.

Reply:
Do you actually know TCL are you just looked it up?  I’ll go back and review this again because I think maybe I see some issues now but I’m not sure. It be really helpful to have some additional test to show what supposed to be possible here. On some of these languages I’m really shooting blind just looking at the examples I have and what the existing rules seem to do.

Reply:
I used to knew it... wrote several programs in Tck/Tk long ago. Now I can hardly remember anything.